### PR TITLE
feat(vats): async serial bridge

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #5287

## Description

This prepares for the split-brain implementation and encapsulates whether downcalls via the `bridge` device are synchronous or asynchronous.

Constrains the use and implementation of `E(bridgeManager).toBridge(dstId, obj)` to ensure downcalls from JS to Golang return a promise for the result, but are only forwarded to Golang one at a time for processing.

### CODE REVIEW

File: packages/vats/src/bridge.js [hunk 1]
```diff
@@ -43,7 +44,8 @@ export function makeBridgeManager(E, D, bridgeDevice) {
 
   const bridgeHandler = Far('bridgeHandler', { inbound: bridgeInbound });
 
-  function callOutbound(dstID, obj) {
+  const withOutboundQueue = makeWithQueue();
+  const toBridge = withOutboundQueue((dstID, obj) => {
```
Comment (mhofman) [inline packages/vats/src/bridge.js:48]: This queue mechanism will still cause all messages sent to the bridge to have their results available within the same crank, right?

File: packages/vats/src/lib-chainStorage.js [hunk 2]
```diff
@@ -49,8 +49,8 @@ export function makeChainStorageRoot(
   function makeChainStorageNode(path, options = {}) {
     const { sequence = false } = options;
     const node = {
-      /** @type {() => VStorageKey} */
-      getStoreKey() {
+      /** @type {() => Promise<VStorageKey>} */
```
Comment (mhofman) [inline packages/vats/src/lib-chainStorage.js:58]: Curious, why are we using `@type` and not `@param` / `@returns` in this file?

File: packages/vats/src/core/chain-behaviors.js [hunk 3]
```diff
@@ -290,19 +289,25 @@ export const startTimerService = async ({
...
+  if (!bridge) {
     console.warn(
       'Running without a bridge device; this is not an actual chain.',
     );
+    bridgeManager.resolve(undefined);
+    return;
   }
-  bridgeManager.resolve(myBridge);
+  const vat = E(loadCriticalVat)('chainStorage');
+  bridgeManager.resolve(E(vat).provideManagerForBridge(bridge));
```
Comment (mhofman) [inline packages/vats/src/core/chain-behaviors.js:310]: If I read this right, this is collocating the bridgeManager in the chainStorage vat to avoid going through the bootstrap vat for operations? Is this a drive-by improvement or is it related to making the bridge async ?

Edit: Oh I see it's because chainStorage used to go through the bridge device directly. Now that it doesn't to avoid incurring extra cost, we need to collocate them.

File: packages/vats/src/vat-chainStorage.js [hunk 4]
```diff
@@ -1,30 +1,55 @@
...
+import { E, Far } from '@endo/far';
+import { makeBridgeManager } from './bridge.js';
 import { makeChainStorageRoot } from './lib-chainStorage.js';
 
 export function buildRootObject(vatPowers) {
+  const { D } = vatPowers;
+  D || Fail`D missing in vatPowers ${Object.keys(vatPowers)}`;
+
+  /** @type {Map<BridgeDevice, import('./types.js').BridgeManager>} */
+  const bridgeToManager = new Map();
```
Comment (mhofman) [inline packages/vats/src/vat-chainStorage.js:12]: I assume this vat isn't durable yet?

File: packages/vats/src/vat-chainStorage.js [hunk 5]
```diff
@@ -48,6 +48,8 @@ export function buildRootObject(vatPowers) {
     return rootNode;
   };
 
+  // We colocate these functions in this vat so that we don't pay extra cranks
+  // to shuffle messages between a chainStorage node and the bridgeManager.
```
Comment (mhofman) [inline packages/vats/src/vat-chainStorage.js:52]: Ha that answers my question!

**Review by mhofman — COMMENTED** [review]
This seem to be a good start but I'm somewhat concerned the result available within same crank semantics may not uncover future complications. I kinda wish we had a simple liveslots / kernel mechanism to split work into another crank, as that would allow simulating the future behavior more faithfully.

**Comment by mhofman** [comment]
> I kinda wish we had a simple liveslots / kernel mechanism to split work into another crank, as that would allow simulating the future behavior more faithfully.

Actually I'm a little confused. I thought we already had a bridge vat. Why can't the chainStorage use the bridge vat instead of using the bridge device directly? That should solve the problem of ensuring async, and remove custom queuing. Then the next step is to change the bridge device interface to not return a result from `callOutbound` but instead provide a promise like mechanism for the bridge to fulfill the result later.

File: packages/vats/src/bridge.js (hunk 1 above)
Comment (michaelfig) [reply packages/vats/src/bridge.js:48]: The results will be available to the bridge vat in the same crank, but will hit the kernel queue before they can be delivered to the caller of this vat.  Does SwingSet run round trips between vats in the same crank?

The queue is only in place to protect the far side of the `bridgeDevice` from receiving more than one request at a time.

File: packages/vats/src/lib-chainStorage.js (hunk 2 above)
Comment (michaelfig) [reply packages/vats/src/lib-chainStorage.js:58]: It was that way when I got here, so I didn't change it.

File: packages/vats/src/vat-chainStorage.js (hunk 4 above)
Comment (michaelfig) [reply packages/vats/src/vat-chainStorage.js:12]: Right.

**Comment by michaelfig** [comment]
> I thought we already had a bridge vat.

Before this PR, the `bridgeManager` was hosted by the bootstrap vat (an antipattern), not a dedicated "bridge vat".

> Why can't the chainStorage use the bridge vat instead of using the bridge device directly?

`chainStorage` (hosted by `vat-chainStorage.js`) used to call the `bridgeManager` from the bootstrap vat, and it was discovered through tracing that all the extra traffic between `chainStorage` and `bridgeManager` seemed like useless overhead.  So, for Pismo, `chainStorage` was modified to use the `bridge` device directly.

So, this PR moves `bridgeManager` out of bootstrap, stops `chainStorage` from using the bridge device directly, and puts them in the same vat to avoid senseless round trips.

> That should solve the problem of ensuring async, and remove custom queuing.

The queue is a distraction.  It's only there to maintain the Golang invariants (serial execution during block evaluation) until we have a comprehensive solution.  All the code surrounding the actual use of the bridge device will have to be rewritten anyway.  This PR was to consolidate that code.

> Then the next step is to change the bridge device interface to not return a result from `callOutbound` but instead provide a promise like mechanism for the bridge to fulfill the result later.

Yup.  But I'm not going to do that until actually splitting the brain.  Baby steps.

File: packages/vats/src/bridge.js [hunk 6]
```diff
@@ -43,7 +44,8 @@ export function makeBridgeManager(E, D, bridgeDevice) {
 
   const bridgeHandler = Far('bridgeHandler', { inbound: bridgeInbound });
 
-  function callOutbound(dstID, obj) {
+  const withOutboundQueue = makeWithQueue();
```
Comment (gibson042) [inline packages/vats/src/bridge.js:47]: These names leave something to be desired.

File: packages/vats/src/bridge.js [hunk 7]
```diff
@@ -53,17 +55,15 @@ export function makeBridgeManager(E, D, bridgeDevice) {
...
   D(bridgeDevice).registerInboundHandler(bridgeHandler);
 
   return Far('bridgeManager', {
-    toBridge(dstID, obj) {
-      return callOutbound(dstID, obj);
-    },
+    toBridge,
     fromBridge(srcID, obj) {
-      return bridgeHandler.inbound(srcID, obj);
+      return E.sendOnly(bridgeHandler).inbound(srcID, obj);
```
Comment (gibson042) [inline packages/vats/src/bridge.js]: Would it make sense to bind `E.sendOnly(bridgeHandler)` rather than regenerating it with every inbound message?

File: packages/vats/src/vat-chainStorage.js [hunk 8]
```diff
@@ -1,30 +1,57 @@
...
-  function makeBridgedChainStorageRoot(bridge, bridgeId, rootPath, options) {
-    const { D } = vatPowers;
-    assert(D, `D missing in vatPowers ${Object.keys(vatPowers)}`);
+  const makeBridgedChainStorageRoot = (bridge, bridgeId, rootPath, options) => {
     // Note that the uniqueness of rootPath is not validated here,
     // and is instead the responsibility of callers.
-    const toStorage = message => D(bridge).callOutbound(bridgeId, message);
+
+    const bridgeManager = provideManagerForBridge(bridge);
+    const toStorage = obj => E(bridgeManager).toBridge(bridgeId, obj);
```
Comment (gibson042) [inline packages/vats/src/vat-chainStorage.js]: Why change from `message` to `obj`?

File: packages/vats/src/vat-chainStorage.js (hunk 5 above)
Comment (gibson042) [reply packages/vats/src/vat-chainStorage.js:52]: I was actually wondering if this is _enough_ colocation. But it looks like any other cross-vat contact is limited to initial setup.

File: packages/vats/test/test-lib-chainStorage.js [hunk 9]
```diff
@@ -8,14 +8,14 @@ test('makeChainStorageRoot', async t => {
   const rootPath = 'root';
   const { rootNode, messages } = makeFakeStorageKit(rootPath);
   t.deepEqual(
-    rootNode.getStoreKey(),
+    await rootNode.getStoreKey(),
```
Comment (gibson042) [inline packages/vats/test/test-lib-chainStorage.js]: This `await` should probably move up out of function arguments.

File: packages/vats/src/bridge.js (hunk 6 above)
Comment (michaelfig) [reply packages/vats/src/bridge.js:47]: How about `outboundQueueWrapper` and `makeQueueWrapper()`?

File: packages/vats/src/bridge.js (hunk 7 above)
Comment (michaelfig) [reply packages/vats/src/bridge.js]: That may be a useful optimisation, it just didn't occur to me.  If it still looks clear enough, I will do it.

File: packages/vats/src/vat-chainStorage.js (hunk 8 above)
Comment (michaelfig) [reply packages/vats/src/vat-chainStorage.js]: Probably habit.  I tend to use `obj` for jsonable values, and `message` for strings.  But that's just me.  I'll put it back.

File: packages/vats/test/test-lib-chainStorage.js (hunk 9 above)
Comment (michaelfig) [reply packages/vats/test/test-lib-chainStorage.js]: Okay, will do that.

File: packages/vats/src/bridge.js (hunk 7 above)
Comment (michaelfig) [reply packages/vats/src/bridge.js]: Oh, did you mean `Function.prototype.bind` or just "a const binding"?  If the latter, then yes, I'll do it.

Comment (mhofman) [reply packages/vats/src/bridge.js]: I believe @erights had qualms about storing the proxy object created by `E`.

File: packages/vats/src/bridge.js (hunk 6 above)
Comment (mhofman) [reply packages/vats/src/bridge.js:47]: I planning on a follow up where we move the queue from the vat side to the host side, so this should all disappear promptly.

File: packages/vats/src/bridge.js (hunk 1 above)
Comment (mhofman) [reply packages/vats/src/bridge.js:48]: Discussed this offline. Plan is to follow up moving the queue to the far side.

**Review by mhofman — APPROVED** [review]
I have a plan to move the queueing into the embedding host side, and make `callOutbound` return `void`, while still keeping the result of `toBridge` as a Promise. Let's land this as-is.
